### DISCUSSION TITLE
Advice for Production Tuning

### DISCUSSION THREAD

**Discussion by keithhackbarth** [Kludex/uvicorn#1874] [Q&A]
Looking for general advice on how to think about tuning for optimized production usage of uvicorn+gunicorn? 

- What things I should be monitoring for?
- How to know if `workers` is configured correctly? (ie, what would be the sign that we have too many)
- Other recommendation settings such as `backlog`, `timeout`, etc.?
- If my server is occasionally throwing a 502 on 1 out of 10,000 queries how to debug?
- Ideal CPU usage (average, max)

**Comment by acdha** [comment]
I was also thinking it would be useful to have more discussion about gunicorn, too. In testing I've found it to be less stable in a containerized environment and slower (I suspect this needs to involve tuning the backlogs in both cases).
